File: apps/web/types/report.ts
import { ChartSpec } from "./chart";

export type ReportStatus = "running" | "completed" | "failed";

export interface Report {
  id: string;
  dataset_id: string;
  title: string;
  goal: string;
  status: ReportStatus;
  error: string | null;
  created_at: string;
  input_tokens: number | null;
  output_tokens: number | null;
  cost_usd: number | null;
  // Lightweight content stats for list/dashboard views.
  section_count: number;
  chart_count: number;
  chart_types: string[];
}

export interface ReportSection {
  title: string;
  narrative: string;
  charts: ChartSpec[];
}

export interface ReportDetail extends Report {
  content: ReportSection[] | null;
}

export type ReportStreamEvent =
  | { type: "report_start"; data: { sections: string[] } }
  | { type: "section_start"; data: { index: number; title: string } }
  | { type: "token"; data: string }
  | { type: "tool_start"; data: { name?: string } }
  | { type: "tool_end"; data: { name?: string } }
  | {
      type: "chart";
      data: { id: string; title: string | null; spec: ChartSpec };
    }
  | { type: "section_end"; data: { index: number } }
  | {
      type: "report_done";
      data: {
        report_id: string;
        cost_usd?: number;
        input_tokens?: number;
        output_tokens?: number;
      };
    }
  | { type: "error"; data: { detail: string } };

File: apps/api/app/schemas/report.py
"""Report request/response DTOs."""
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.report import ReportStatus


class ReportCreate(BaseModel):
    dataset_id: uuid.UUID
    # Either provide a free-text goal, or a suggestion_id to build the report
    # from an AI suggestion (its question/rationale become the goal).
    goal: str | None = Field(default=None, description="What the report should cover.")
    title: str | None = Field(default=None, max_length=255)
    suggestion_id: uuid.UUID | None = None


class ReportUpdate(BaseModel):
    title: str = Field(min_length=1, max_length=255)


class ReportRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    dataset_id: uuid.UUID
    title: str
    goal: str
    status: ReportStatus
    error: str | None
    created_at: datetime
    # Token usage + priced cost. Null for reports made before cost tracking.
    input_tokens: int | None = None
    output_tokens: int | None = None
    cost_usd: float | None = None
    archived: bool = False
    # Lightweight content stats for list views (computed from `content`), so the
    # dashboard can show how many sections/charts a report has and which chart
    # types it uses — without shipping the full report body.
    section_count: int = 0
    chart_count: int = 0
    chart_types: list[str] = []


class ReportDetail(ReportRead):
    # List of sections: {title, narrative, charts: [vega_spec, ...]}
    content: list | None
